### PR TITLE
Use local auth.json file when available

### PR DESCRIPTION
**Pull request by danez**
This ensures renovate bot reads a local `auth.json` file if it is available next to `composer.json` and uses it as the default before modyfing it.

I wasn't sure if values from a local `auth.json` should be overwritten or not. I added a test that tests that the values are correctly overwritten by renovate bot, but I'm happy to change it. I just went for overwritting for now, as it requires less modifications to the code. 
Let me know what you think the expected behaviour should be.

Does NOT fix issue in #5616.

### CODE REVIEW

File: lib/manager/composer/artifacts.spec.ts [hunk 1]
```diff
@@ -85,6 +94,80 @@ describe('.updateArtifacts()', () => {
       })
     ).toBeNull();
     expect(execSnapshots).toMatchSnapshot();
+    expect(fs.outputFile.mock.calls[1][1]).toMatchSnapshot();
```
Comment (danez) [inline lib/manager/composer/artifacts.spec.ts]: This test even though it's name is `uses hostRules to write auth.json` never asserted what is actually written to `auth.json`. So I added one more expectation.

**Review by viceice — COMMENTED** [review]
@danez Those `auth.json` files should never been commited to a repository. 😱  They contain unprotected authentication data.

@rarkins Are we sure to support this scenario?

**Comment by rarkins** [comment]
@viceice what people do in their own homes..

Sometimes people also commit `.npmrc` files with auth tokens too, and we support that. Also I don't think there are any *tradeoffs* in supporting this, so that's another positive point.

I thought though we should maybe look for this file and import it into `hostRules` early in the extract phase so that lookups can succeed too. Then the `auth.json` can be composed entirely from `hostRules` in the artifacts stage.

**Comment by hongaar** [comment]
> I'm not sure if this actually fixes the issues that @hongaar reported in #5616.

@danez we don't use a custom `auth.json` file in our use-case, so I don't think the proposed changes would change anything for us.

**Comment by danez** [comment]
> I thought though we should maybe look for this file and import it into `hostRules` early in the extract phase so that lookups can succeed too. Then the `auth.json` can be composed entirely from `hostRules` in the artifacts stage.

Do i understand this correctly, that i should add the logic for reading the auth.json file to `composer/extract.ts`? And in this file should I simply add entries to `hostRules.add()` or return the hostrules with the result and then do `hostRules.add()` wherever the extract function is called? The second approach seems more generic to me, so that also other managers could simple return hostRules, but not sure if this is desired.

**Comment by rarkins** [comment]
@danez I think looking for it and reading it during the extract phase makes sense, similar to how we do for `.npmrc`. 

The second approach does seem more generic, although not sure if it reduces any code? i.e. returning a `hostRules` array or just calling `hostRules.add()` seem like equivalent work. Maybe just call `.add()` directly for now because this would be the first manager doing so and we avoid prematurely optimizing.

File: lib/manager/composer/artifacts.ts [hunk 2]
```diff
@@ -40,10 +40,14 @@ export async function updateArtifacts({
     const localPackageFileName = upath.join(config.localDir, packageFileName);
     await fs.outputFile(localPackageFileName, newPackageFileContent);
     const localLockFileName = upath.join(config.localDir, lockFileName);
+    const localAuthFileName = upath.join(cwd, 'auth.json');
     if (config.isLockFileMaintenance) {
       await fs.remove(localLockFileName);
     }
-    const authJson = {};
+    let authJson = {};
+    if (await fs.pathExists(localAuthFileName)) {
```
Comment (bytestream) [inline lib/manager/composer/artifacts.ts:48]: This should also check `COMPOSER_HOME`
> The third way if you want to pre-configure it is via an auth.json file located in your COMPOSER_HOME or besides your composer.json.

Ref;
* https://getcomposer.org/doc/articles/http-basic-authentication.md
* https://getcomposer.org/doc/03-cli.md#composer-home

**Comment by bytestream** [comment]
> @danez Those auth.json files should never been commited to a repository. 😱 They contain unprotected authentication data.

No need to commit to the repository. In GitLab you can store it in an encrypted CI variable and create the `auth.json` in the CI environment:
```
- mkdir -p ~/.composer/ && echo "$COMPOSER_AUTH_CONFIG" > ~/.composer/auth.json
```

**Comment by danez** [comment]
We are not only doing that for ci but also for developer machines, so not everyone has to login themselves.

**Comment by danez** [comment]
> @danez I think looking for it and reading it during the extract phase makes sense, similar to how we do for `.npmrc`.
> 
> The second approach does seem more generic, although not sure if it reduces any code? i.e. returning a `hostRules` array or just calling `hostRules.add()` seem like equivalent work. Maybe just call `.add()` directly for now because this would be the first manager doing so and we avoid prematurely optimizing.

I tried to do this, but stumbled upon on major question. 

I tried adding code that extracts all possible credentials from auth.json (github/gitlab-oauth, gitlab-token, plain http-basic) and adds them to the hostrules. So far so good, that works. In `updateArtifacts` the github and gitlab tokens would be set correctly from the new host-rules, but http basic is currently only written to auth.json if the url is set as registry url, which in our case it is not, because we set the http basic auth options for our git repository and not a registry.

To make it actually work I would need to write ALL credentials in the host-rules to the generated `auth.json` in `updateArtifacts`, because I cannot differentiate which host-rules actually came from `auth.json` (Or maybe there is a way to 'tag' credentials when adding them to host-rules, so it would only write whatever was read from auth.json? But I don't see anything in the type `HostRule` that allows to do this.)

This seems wrong to me to simply dump everything in host-rules to auth.json. Also if repos are not reset between runs, renovate would end up rereading everything again what it previously wrote?

Another option would be to read `auth.json` in extract step, add it to host-rules, but then reading it again in `updateArtifacts` and do the logic that is currently here in this PR.

So not really sure how to solve this.

**Comment by rarkins** [comment]
@danez you could consider doing it similar to Bundler: https://github.com/renovatebot/renovate/blob/51a43d5e41e0727b618553a688ff554898677d66/lib/manager/bundler/artifacts.ts#L120-L124

i.e. all hostRules have hostType=composer configured and then you'd write out only those rules to `auth.json` instead of all.

File: lib/manager/composer/artifacts.ts (hunk 2 above)
Comment (danez) [reply lib/manager/composer/artifacts.ts]: I haven't done this change in this PR, because the main reason I created the PR right now was to not make renovate not overwrite a committed `auth.json` file. The `auth.json` file in `COMPOSER_HOME` would still work as usual for composer because renovate does not do anything to it. This would also be on the host were renovate runs and I wonder if anyone creates an `auth.json` file there. I would rather wait for a specific use case before doing it.

Comment (bytestream) [reply lib/manager/composer/artifacts.ts]: I use `COMPOSE_HOME` that's why I mentioned it hah:
> `mkdir -p ~/.composer/ && echo "$COMPOSER_AUTH_CONFIG" > ~/.composer/auth.json`

Your suggestion is to output the file into the build / current working directory?

**Comment by danez** [comment]
I made the modifications and moved the logic into `extract.ts`. :)

File: lib/manager/composer/artifacts.ts (hunk 2 above)
Comment (danez) [reply lib/manager/composer/artifacts.ts]: Right, but afaics this is not affected at all by this PR and will continue to work.

**Comment by danez** [comment]
The more I think about this while discussing with @bytestream above, I think this is an incomplete solution. 🤔

The authentication options, that I implemented here to be read from `projectpath/auth.json` could also be in:
- `COMPOSER_HOME/auth.json` (Usually `~/.composer/auth.json`)
- `COMPOSER_HOME/config.json` (Usually `~/.composer/config.json`)
- `projectpath/composer.json` (in the `config` property)

And I guess for git repos that need authentication, renovate bot internally needs to know the credentials to be able to access these repos to check for new tags? I haven't thought about this earlier, because in our case `composer install` was simply failing because renovate was overwriting the local `auth.json`. So if I understand now correctly the initial approach (without hostRules, simply amending an existing `auth.json`) wouldn't have worked because renovate wouldn't have the credentials to access the git repo.

Implementing all the different config files and duplicating the merging logic from composer in renovate seems prune for error. Composer has a command to get the current config for a repo (`composer config --list`) but unfortunately it is not really machine readable.

```
...
[github-oauth.github.com] abcdefghijk
[http-basic.git.domain.net.username] user
[http-basic.git.domain.net.password] password
```

What would you think if I try to get composer to add a json output format to the command mentioned above and once it is available simply calling this instead of reading files? This roundtrip might take a while though and renovate would require this new version of composer-.

File: lib/manager/composer/artifacts.ts (hunk 2 above)
Comment (bytestream) [reply lib/manager/composer/artifacts.ts]: Ah sorry, I might have completely misunderstood this PR. You're right, global `auth.json` works just fine.

**Comment by rarkins** [comment]
This sounds good/important if you're running it locally on your developer machine, but I don't think it's a good idea to run Renovate in production on a shared machine. Ideally Renovate's should be configured either by its own config or from repository contents and not by hard-to-understand locations like $COMPOSER_HOME etc

**Comment by danez** [comment]
Yeah okay makes sense. Then this PR solves the issue with renovate overwriting committed auth.json files.
I tested in the dev container (had to install PHP and composer in the container).

Without the change I get this error in the logs:
```
Command failed: composer update phpunit/phpunit --with-dependencies --ignore-platform-reqs --no-ansi --no-interaction --no-scripts --no-autoloader
Loading composer repositories with package information

                                                                                                                                                                                                   
  [RuntimeException]                                                                                                                                                                               
  Failed to execute git clone --mirror 'https://git.domain.net/scm/php/mongodb-odm-light.git' '/tmp/renovate/cache/others/composer/vcs/https---git.domain.net-scm-php-mongodb-odm-light.git/'  
                                                                                                                                                                                                   
  Cloning into bare repository '/tmp/renovate/cache/others/composer/vcs/https---git.domain.net-scm-php-mongodb-odm-light.git'...                                                                 
  Remote-Containers CLI: RPC pipe not configured. Message: {"command":"fill","stdin":"protocol=https\nhost=git.domain.net\n"}                                                                    
  Remote-Containers CLI:***@git.domain.net': \n"}                                                                                                                                                
  fatal: Authentication failed for 'https://git.domain.net/scm/php/mongodb-odm-light.git/'                                                                                                       
```

After applying this PR it works correctly and packagist dependencies are successfully updated.

I ran into some more issues with "custom vcs repositories" in composer.json, but this is unrelated to this. Will create another issue. :)

**Comment by rarkins** [comment]
@danez could you summarize for me:
- what this PR now achieves
- how/when it does it in the process
- any limitations or concerns
- future work?

**Comment by danez** [comment]
> * what this PR now achieves

The PR changes the composer manager to read `auth.json` and add all the stored credentials in there to host-rules (with `hostType: composer` for http basic credentials, tokens for github/gitlab have their corresponding hostType). When generating `auth.json` later it will re-add this http basic credentials.

> * how/when it does it in the process
The extraction of the credentials happens in the extract phase (https://github.com/renovatebot/renovate/pull/5653/files?file-filters%5B%5D=.ts#diff-4f587f44fa207e89b9941f31686a7650R192) at the beginning of `extractPackageFile`

The reinsertion of the credentials happens in `updateArtifacts` (https://github.com/renovatebot/renovate/pull/5653/files?file-filters%5B%5D=.ts#diff-c28119bdd5353abd6f3cb14c1ed62b6bR87-R98) The diff here is a little bit hard to read, because I extracted the code that created auth.json into its own function.

> * any limitations or concerns

One thing I am not sure about is how multiple hostRules are handled. For example what if a github token is configured in the renovate config and one in a committed `auth.json`. The code will add the one from `auth.json` to hostRules, but only one is read from the hostRules later. Not sure which one though.

> * future work?

I don't see any, this PR handles exactly the one problem when `auth.json` files are committed to vcs.


P.S sorry for the long wait for my answer :)

File: lib/manager/composer/artifacts.ts [hunk 3]
```diff
@@ -48,62 +126,12 @@ export async function updateArtifacts({
...
-          { registryUrls: config.registryUrls },
-          'Non-array composer registryUrls'
-        );
-      }
-    } catch (err) /* istanbul ignore next */ {
-      logger.warn({ err }, 'Error setting registryUrls auth for composer');
-    }
+
+    const authJson = generateAuthJson(config);
     if (authJson) {
```
Comment (danez) [inline lib/manager/composer/artifacts.ts:131]: This was always `true` as authJson was always an object. The new code will return null and therefore not write an empty auth.json files.

**Comment by rarkins** [comment]
Blocked by #6907

**Comment by rarkins** [comment]
@danez I would like to merge #6932 first and then this PR can hopefully be reduced to just the extract part. Are you able to review or test it?

File: lib/manager/composer/extract.ts [hunk 4]
```diff
@@ -84,6 +117,61 @@ function parseRepositories(
...
+  const localAuthFileName = upath.join(path, 'auth.json');
+  if (!(await pathExists(localAuthFileName))) {
+    return;
+  }
+
+  const authJson: ComposerAuthConfig = await readJson(localAuthFileName);
+
+  if (authJson['github-oauth']) {
+    Object.entries(authJson['github-oauth']).forEach(([domainName, token]) => {
+      hostRules.add({
```
Comment (rarkins) [inline lib/manager/composer/extract.ts:133]: These hostRules.add should all include hostType=packagist

**Comment by danez** [comment]
I can do that. Will check.

But as the other PR is removing the writing of `auth.json`, it cometelly solves the problem i had with a commited `auth.json`.

Unless you think the importing of credentials is a good feature, i would say we can close this if the other PR is merged.

**Comment by rarkins** [comment]
I like the work you did to convert any `auth.json` credentials in the repo into `hostRules`, although I guess it's quite an antipattern to commit credentials to repos these days anyway. 

I've taken a local copy of your branch so feel free to close the PR if your own problems will be resolved!
